stablehlo/transforms/ChloDecompositionPatternsMath.td

```diff
@@ -384,3 +384,77 @@ def : Pat<(CHLO_AsinOp NonComplexElementType:$x),
           (StableHLO_MulOp
             (StableHLO_SubtractOp $one, $x),
             (StableHLO_AddOp $one, $x))))))>;
+
+// Inverse hyperbolic sine on complex input:
+//
+//     asinh(z) = -I * asin(I * z)
+//
+def : Pat<(CHLO_AsinhOp ComplexElementType:$z),
+  (StableHLO_ComplexOp
+    (StableHLO_ImagOp
+      (CHLO_AsinOp:$w
+        (StableHLO_ComplexOp
+          (StableHLO_NegOp
+            (StableHLO_ImagOp $z)),
+          (StableHLO_RealOp $z)))),
+    (StableHLO_NegOp
+      (StableHLO_RealOp $w)))>;
+
+// Inverse hyperbolic sine on real input:
+//
+//       asinh(x) = log(x + hypot(1, x))
+//
+//     This algorithm is based on the StableHLO v1.0 function CHLO_AsinhOp.
+//
+//     To avoid overflow in x * x, we use
+//
+//       asinh(x) = log(2) + log(x)
+//
+//     when abs(x) > sqrt(max),
+//
+//     To avoid underflow in 1 + x * x, we'll define z = hypot(1, x) and
+//     write
+//
+//       log(x + z)
+//       = log(((x + z) * (1 + z)) / (1 + z))
+//       [ z ** 2 = 1 + x ** 2]
+//       = log((x + x * z + z + 1 + x**2) / (1 + z))
+//       = log(1 + x + x ** 2 / (1 + z))
+//       = log1p(x + x ** 2 / (1 + hypot(1, x)))
+//
+//     that is, for non-negative x, we have
+//
+//       asinh(x) = log1p(x + x ** 2 / (1 + hypot(1, x)))
+//
+//     It turns out, this is accurate for all abs(x) < sqrt(max).
+//
+//     For x < 0, we'll use
+//
+//       asinh(x) = -asinh(-x)
+//
+//
+def : Pat<(CHLO_AsinhOp NonComplexElementType:$x),
+  (StableHLO_MulOp
+    (StableHLO_SignOp $x),
+    (StableHLO_SelectOp
+      (StableHLO_CompareOp
+       (StableHLO_AbsOp:$ax $x),
+       (StableHLO_MulOp
```

Comment:
Why this MulOp by 1?

Reply:
Thanks for noticing this!

The origin of "MulOp by 1" is https://github.com/pearu/functional_algorithms/blame/cefa141d5ab2775ed0ece7a4d3892f1d59ced4c6/functional_algorithms/algorithms.py#L373
that is, it comes from parameterization of approximation boundaries. In this particular case, `1` turned to be optimal.

The fix is on the way.

Reply:
Fixed.